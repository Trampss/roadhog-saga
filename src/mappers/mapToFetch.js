import { select } from 'redux-saga/effects'
import { getOptions, getURL } from '../selectors'
import { addPathParams, addQueryParams } from '../url-pattern'

export default action => function* (inputs = {}) {
  // action pattern is `<METHOD_NAME>_<RESOURCE_NAME>`
  const [methodName, name] = action.split(/_(.+)/)

  // create a fetch options object
  let options = yield select(getOptions(name)(methodName))
  options = { ...options, body: inputs.body }

  // create a fetch url
  let url = yield select(getURL(name)(methodName))
  url = addPathParams(url)(inputs.path)
  url = addQueryParams(url)(inputs.query)

  return [url, options]
}
