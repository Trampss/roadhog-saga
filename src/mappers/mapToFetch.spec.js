/* eslint-env jest */
import tester from 'trampss-redux-saga-tester'
import mapToFetch from './mapToFetch'

jest.mock('../url-pattern', () => ({
  addPathParams: url => pathParams => `url:${url}//pathPrams:${JSON.stringify(pathParams)}`,
  addQueryParams: url => queryParams => `url:${url}//queryParams:${JSON.stringify(queryParams)}`,
}))

describe('mapToFetch', () => {
  const test = tester(mapToFetch('an_action'))
  const mocks = {
    select: [
      () => ({ some: 'options' }),
      () => 'http://an-url',
    ],
  }

  it('should set all params', () => {
    expect(
      test({
        body: { a: 'body', to: 'send' },
        path: ['an', 'great', 'array'],
        query: { an: 'object', that: 'rocks' },
      })(mocks),
    ).toMatchSnapshot()
  })

  it('should set minimum params', () => {
    expect(
      test()(mocks),
    ).toMatchSnapshot()
  })
})
