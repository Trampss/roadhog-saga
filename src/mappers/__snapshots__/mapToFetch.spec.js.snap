exports[`mapToFetch should set all params 1`] = `
Array [
  Object {
    "raw": Object {
      "done": false,
      "value": Object {
        "@@redux-saga/IO": true,
        "SELECT": Object {
          "args": Array [],
          "selector": [Function],
        },
      },
    },
    "value": Object {
      "some": "options",
    },
  },
  Object {
    "raw": Object {
      "done": false,
      "value": Object {
        "@@redux-saga/IO": true,
        "SELECT": Object {
          "args": Array [],
          "selector": [Function],
        },
      },
    },
    "value": "http://an-url",
  },
  Object {
    "done": true,
    "value": Array [
      "url:url:http://an-url//pathPrams:[\"an\",\"great\",\"array\"]//queryParams:{\"an\":\"object\",\"that\":\"rocks\"}",
      Object {
        "body": Object {
          "a": "body",
          "to": "send",
        },
        "some": "options",
      },
    ],
  },
]
`;

exports[`mapToFetch should set minimum params 1`] = `
Array [
  Object {
    "raw": Object {
      "done": false,
      "value": Object {
        "@@redux-saga/IO": true,
        "SELECT": Object {
          "args": Array [],
          "selector": [Function],
        },
      },
    },
    "value": Object {
      "some": "options",
    },
  },
  Object {
    "raw": Object {
      "done": false,
      "value": Object {
        "@@redux-saga/IO": true,
        "SELECT": Object {
          "args": Array [],
          "selector": [Function],
        },
      },
    },
    "value": "http://an-url",
  },
  Object {
    "done": true,
    "value": Array [
      "url:url:http://an-url//pathPrams:undefined//queryParams:undefined",
      Object {
        "body": undefined,
        "some": "options",
      },
    ],
  },
]
`;
